src/ui/castleinternaljoystickswindows.pas

```diff
@@ -215,75 +245,90 @@ procedure TWindowsJoysticksBackend.Poll(const List: TJoystickList;
+                EventContainer.OnAxisMove(Joystick, j, _value);
+            Joystick.State.Axis[ a ] := _value;
+          end;
+
+          FillChar( Joystick.State.Axis[ JOY_POVX ], 8, 0 );
+          if ( Joystick.Info.Caps and JOY_HAS_POV > 0 ) and ( state.dwPOV and $FFFF <> $FFFF ) then
+          begin
+            _value := Sin( DegToRad(state.dwPOV and $FFFF / 100.0) );
+            if Joystick.State.Axis[ JOY_POVX ] <> _value then
+              if Assigned(EventContainer.OnAxisMove) then
+                EventContainer.OnAxisMove(Joystick, JOY_POVX, _value);
+            Joystick.State.Axis[ JOY_POVX ] := _value;
+
+            _value := -Cos( DegToRad(state.dwPOV and $FFFF / 100.0 ) );
+            //_value := -_value;
+            if Joystick.State.Axis[ JOY_POVY ] <> _value then
+              if Assigned(EventContainer.OnAxisMove) then
+                EventContainer.OnAxisMove(Joystick, JOY_POVY, _value);
+            Joystick.State.Axis[ JOY_POVY ] := _value;
+          end;
+
+          for j := 0 to Joystick.Info.Count.Buttons - 1 do
+          begin
+            btn := state.wButtons and ( 1 shl j );
+            if ( Joystick.State.BtnDown[ j ] ) and ( btn = 0 ) then
+            begin
+              Joystick.State.BtnPress[ j ] := False;
+              if Assigned(EventContainer.OnButtonUp) then EventContainer.OnButtonUp(Joystick, j);
+              Joystick.State.BtnCanPress[ j ] := True;
+            end;
+
+            if ( Joystick.State.BtnCanPress[ j ] ) and ( not Joystick.State.BtnDown[ j ] ) and ( btn <> 0 ) then
+            begin
+              Joystick.State.BtnPress   [ j ] := True;
+              if Assigned(EventContainer.OnButtonPress) then EventContainer.OnButtonPress(Joystick, j);
+              Joystick.State.BtnCanPress[ j ] := False;
+            end;
+            Joystick.State.BtnDown[ j ] := btn <> 0;
+            Joystick.State.BtnUp[ j ] := btn = 0;
+            if Assigned(EventContainer.OnButtonDown) and (btn <> 0) then EventContainer.OnButtonDown(Joystick, j);
+          end;
         end;
-
-        if ( Joystick.State.BtnCanPress[ j ] ) and ( not Joystick.State.BtnDown[ j ] ) and ( btn <> 0 ) then
+      JOYERR_UNPLUGGED:
         begin
-          Joystick.State.BtnPress   [ j ] := True;
-          if Assigned(EventContainer.OnButtonPress) then EventContainer.OnButtonPress(Joystick, j);
-          Joystick.State.BtnCanPress[ j ] := False;
+          WritelnLog('Joystick %s was disconnected', [Joystick.Info.Name]);
+          JoystickHadBeenDisconnected := true;
         end;
-        Joystick.State.BtnDown[ j ] := btn <> 0;
-        Joystick.State.BtnUp[ j ] := btn = 0;
-        if Assigned(EventContainer.OnButtonDown) and (btn <> 0) then EventContainer.OnButtonDown(Joystick, j);
-      end;
+      JOYERR_PARMS:
+        WritelnLog('Joystick %s parameters are no longer valid.', [Joystick.Info.Name]);
+      else
+        WritelnLog('Joystick %s error %d', [Joystick.Info.Name, JoyError]);
```

Comment:
This should be `WritelnWarning` (unless it's something that can routinely occur during normal program execution and we consciously know that it can be ignored, and doesn't indicate any error in our code).